### PR TITLE
Enable alerting routes per team

### LINKED ISSUE #148: Enable alerts to go to specific team channels
Product teams need freedom to send alerts to different slack channels.

Monitoring-satellite should be able to decide which slack channel to send alerts based on a label provided by the alert!

### PR DESCRIPTION
**Pull request by ArthurSens**
## Description

This PR enables alerts to be routed to team-specific slack channels or pagerduty schedules, depending on whether the alert has a label `team` attached to it or not, and also by its value.

It was implemented in a way that teams can decide when to start using team specific slack channels or pagerduty schedules. Following the rules below:

```mermaid
graph TB;
  Prometheus-- Send alert -->Alertmanager;
  Alertmanager-- severity=warning --> slackWarningReceiver([Slack channel 'warning catches all']);
  Alertmanager-- severity=info--> slackInfoReceiver([Slack channel 'info catches all']);
  Alertmanager-- severity=critical--> hasTeamLabel{Does it have a team label?};
  hasTeamLabel-- No --> genericPDReceiver([Route to company-wide Pagerduty schedule]);
  hasTeamLabel-- Yes --> hasSpecificPDKey{Does it have a PagerDuty routing key configured?};
  hasSpecificPDKey-- No --> teamSlackChannel([Route to team-specific slack channel]);
  hasSpecificPDKey-- Yes --> teamPDSchedule([Route to team-specific PagerDuty schedule]);
```

## Related Issue(s)

Fixes #148 

## How to test

You can play around with different configurations under `hack/generate.sh` and see the generation results under `monitoring-satellite/manifests/alertmanager/secret.yaml`

Copy the generated secret into https://www.prometheus.io/webtools/alerting/routing-tree-editor/, and test different alert to check if it is going to the expected receiver

### CODE REVIEW

File: hack/generate.sh
Comment (ArthurSens) [inline hack/generate.sh]: This is an example configuration that we'll have to add ArgoCD apps to enable team-specific alerting routes. 

They can choose between slack or pagerduty, and the company-wide routes will be used if a team doesn't have such configuration.

File: addons/alerting.libsonnet
Comment (mads-hartmann) [inline addons/alerting.libsonnet]: This would be a little bit more readable if we did something like

```jsonnet
local globalPagerDutyRoutingKey = if std.objectHas(config.alerting, 'pagerdutyRoutingKey')
  then config.alerting.pagerdutyRoutingKey
  else ""

local IDEPagerDutyRoutingKey = if std.objectHas(config.alerting, 'IDE') && std.objectHas(config.alerting.IDE, 'pagerdutyRoutingKey') 
  then config.alerting.IDE.pagerdutyRoutingKey
  else globalPagerDutyRoutingKey

local IDECriticalReceiver =
  if std.length(IDEPagerDutyRoutingKey) > 0
  then 
    |||
        pagerduty_configs:
          - send_resolved: true
            routing_key: '%(pagerdutyRoutingKey)s'
    ||| % {
        pagerdutyRoutingKey: IDEPagerDutyRoutingKey
    }
  else
    // Slack goes here
```

File: addons/alerting.libsonnet
Comment (mads-hartmann) [inline addons/alerting.libsonnet]: Would it be possible to extract this to a local variable too and then reuse it for all of the receivers? Something like

```jsonnet
local slackConfig = |||
  slack_configs:
    - send_resolved: true
      api_url: %(slackWebhookUrlCritical)s
      channel: '%(slackChannelPrefix)s_critical'
      title: '[{{ .Status | toUpper }}{{ if eq .Status "firing" }}{{ end }}] %(clusterName)s Monitoring'
      text: |
        {{ range .Alerts }}
        **Please take immediate action!**
        *Cluster:* {{ .Labels.cluster }}
        *Alert:* {{ .Labels.alertname }}
        *Description:* {{ .Annotations.description }}
        {{ end }}
      actions:
      - type: button
        text: 'Runbook :book:'
        url: '{{ .CommonAnnotations.runbook_url }}'
|||

local IDECriticalReceiver = if .. 
  then ... 
  else slackConfig % {
    // all params goes here
  }
```

**Review by mads-hartmann — COMMENTED** [review]
This is excellent and a good 🛹 

I left a few suggestions for how we might be able to make the jsonnet a bit more readable.

File: addons/alerting.libsonnet
Comment (ArthurSens) [reply addons/alerting.libsonnet]: Probably 🙂

File: addons/alerting.libsonnet
Comment (ArthurSens) [reply addons/alerting.libsonnet]: This actually doesn't work because it makes the global pagerduty key have preference over team-specific slack webhook 😕

Comment (mads-hartmann) [reply addons/alerting.libsonnet]: Ah right, reading the code it looks like that was the case before too. Can we just simplify it and say

1. If there is an IDE pagerdutyRoutingKey then use PD
2. Otherwise use Slack (using the team slack, falling back to global slack)

Something like

```jsonnet
local IDEPagerDutyRoutingKey = if std.objectHas(config.alerting, 'IDE') && std.objectHas(config.alerting.IDE, 'pagerdutyRoutingKey') 
  then config.alerting.IDE.pagerdutyRoutingKey
  else ""

local IDECriticalReceiver =
  if std.length(IDEPagerDutyRoutingKey) > 0
  then 
    |||
        pagerduty_configs:
          - send_resolved: true
            routing_key: '%(pagerdutyRoutingKey)s'
    ||| % {
        pagerdutyRoutingKey: IDEPagerDutyRoutingKey
    }
  else
    // Slack goes here
```

Comment (mads-hartmann) [reply addons/alerting.libsonnet]: Ah wait, I finally see the subtle distinction that if there wasn't an config.alerting.IDE object at all it would use the global PD key ☺️

I'm still leaning towards going with the suggestion before though (that if you configure an alert for a team but you don't have any alerting config for the team, then it just goes to slack) just to keep the logic simple

Comment (ArthurSens) [reply addons/alerting.libsonnet]: Still doesn't work like that :grimacing:

I tried to explain it here https://www.loom.com/share/6d4ccbed310b4b4b978d93ba52130a9b

Comment (mads-hartmann) [reply addons/alerting.libsonnet]: @ArthurSens Thanks for the loom, looking at the code you have open it's missing a minor change I made in the comment above (that the "else" branch returns "" instead of the global routing key) ☺️ I'm about to head out but if I'm still misunderstanding something I will try to be online tonight and perhaps we can have a quick sync call instead ☺️

**Review by mads-hartmann — APPROVED** [review]
@ArthurSens As my comments are not about functional changes but just tidying up the code a little bit to make it shorter and the logic easier to follow I'm going to approve this and you can just merge if you want and we can follow up with another PR - I should have done that initially, didn't mean to stall thing ☺️

**Comment by ArthurSens** [comment]
> @ArthurSens As my comments are not about functional changes but just tidying up the code a little bit to make it shorter and the logic easier to follow I'm going to approve this and you can just merge if you want and we can follow up with another PR - I should have done that initially, didn't mean to stall thing ☺️

No worries, I very much appreciate suggestions to make the code cleaner 
I believe I'm almost there, so we can wait until the next commit before merging 🙂

**Comment by ArthurSens** [comment]
Alright Mads! A couple of changes since your last review:

* Made the code more readable following your suggestions
* Slack messages seem to break if we use a slack webhook that is not coupled with the correct channel. So now I'm also asking for consumers to inform the slack-channel. An assertion was added to make sure consumers don't forget to inform both channel and webhook

**Review by mads-hartmann — APPROVED** [review]
Awesome, looks great!

**Pull request merged**
